File: frontend/types/action-center.ts
// 行动任务中心类型定义（对齐后端 app/schemas/action.py）
// 契约枚举按后端以 str + 注释形式声明

/** 行动类型枚举 */
export type ActionType =
  | "read_article"
  | "finish_course"
  | "resume_revise"
  | "mock_interview"
  | "real_apply"
  | "get_offer"
  | "custom";

/** 行动状态枚举 */
export type ActionStatus = "PENDING" | "DONE" | "EXPIRED" | "CANCELED";

/** 连击状态枚举 */
export type StreakStatus = "ACTIVE" | "BROKEN" | "NEVER";

/** 创建行动项请求（user_id 由登录态 token 推断） */
export interface ActionCreateRequest {
  action_type: ActionType;
  title: string;
  /** 计划完成日期；格式 年-月-日（如 2026-01-15） */
  due_date: string;
  /** 来源决策分析 ID（决策中心联动） */
  source_decision_id?: number | null;
  /** 备注（契约字段，当前创建时忽略） */
  note?: string | null;
  /** 业务方扩展（契约字段，当前创建时忽略） */
  biz_fields?: Record<string, unknown>;
}

/** 更新行动项请求（部分更新，均可选） */
export interface ActionUpdateRequest {
  title?: string | null;
  due_date?: string | null;
  status?: ActionStatus | null;
  note?: string | null;
}

/** 行动打卡请求 */
export interface CheckinRequest {
  action_id: number;
  /** 打卡时间 ISO 字符串 */
  completed_at: string;
  evidence_url?: string | null;
  note?: string | null;
}

/** 连续天数统计 VO */
export interface StreakVO {
  user_id: string;
  current_streak_days: number;
  longest_streak_days: number;
  last_checkin_date: string | null;
  streak_status: StreakStatus;
}

/** 行动项 VO */
export interface ActionVO {
  id: number;
  user_id: string;
  action_type: ActionType;
  title: string;
  due_date: string;
  source_decision_id: number | null;
  weight: number;
  status: ActionStatus;
  created_time: string;
}

/** 行动列表 VO */
export interface ActionListResponse {
  items: ActionVO[];
  total: number;
}

/** 打卡记录 VO */
export interface CheckinVO {
  id: number;
  action_id: number;
  user_id: string;
  completed_at: string;
  evidence_url: string | null;
  note: string | null;
  biz_req_no: string;
}

/** 打卡历史列表 VO */
export interface CheckinListResponse {
  items: CheckinVO[];
  total: number;
}

/** 行动权重 VO */
export interface ActionWeightVO {
  id: number;
  action_type: ActionType;
  weight: number;
  weight_label: string;
  enabled: boolean;
}

/** 行动权重列表 VO */
export interface ActionWeightListResponse {
  items: ActionWeightVO[];
  total: number;
}

File: backend/app/models/action_center.py
"""行动任务中心契约模型（系统设计 §4.2.1 ~ §4.2.4）。

MVP 方案 C：契约先行、实现延后 — 仅落库建表，业务逻辑后续实现。
"""
from datetime import date, datetime
from uuid import UUID

from sqlalchemy import (
    BigInteger,
    Boolean,
    Date,
    DateTime,
    Index,
    Integer,
    String,
    UniqueConstraint,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base
from app.models.base import GUID, BigIntPK, ContractAuditMixin


class DailyAction(ContractAuditMixin, Base):
    """每日行动任务（t_action）。

    状态机：PENDING → DONE / EXPIRED / CANCELED。
    唯一约束：同一用户同一天同一类型只能存在一条行动（UK_user_id_action_type_due_date）。
    """
    __tablename__ = "t_action"

    id: Mapped[int] = mapped_column(BigIntPK, primary_key=True, autoincrement=True)
    user_id: Mapped[UUID] = mapped_column(GUID(), nullable=False)
    action_type: Mapped[str] = mapped_column(String(20), nullable=False)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    due_date: Mapped[date] = mapped_column(Date, nullable=False)
    source_decision_id: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    weight: Mapped[int] = mapped_column(
        Integer, nullable=False, default=1, server_default=text("1")
    )
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="PENDING", server_default=text("'PENDING'")
    )
    # PENDING / DONE / EXPIRED / CANCELED
    biz_req_no: Mapped[str | None] = mapped_column(String(64), nullable=True)

    __table_args__ = (
        UniqueConstraint(
            "user_id", "action_type", "due_date",
            name="uk_action_user_id_action_type_due_date",
        ),
        Index("idx_action_user_id_due_date", "user_id", "due_date"),
        Index("idx_action_user_id_status", "user_id", "status"),
    )


class ActionCheckin(ContractAuditMixin, Base):
    """行动打卡记录（t_action_checkin）。"""
    __tablename__ = "t_action_checkin"

    id: Mapped[int] = mapped_column(BigIntPK, primary_key=True, autoincrement=True)
    action_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    user_id: Mapped[UUID] = mapped_column(GUID(), nullable=False)
    completed_at: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    evidence_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    note: Mapped[str | None] = mapped_column(String(500), nullable=True)
    biz_req_no: Mapped[str] = mapped_column(String(64), nullable=False)

    __table_args__ = (
        UniqueConstraint("biz_req_no", name="uk_action_checkin_biz_req_no"),
        Index("idx_action_checkin_user_id_completed_at", "user_id", "completed_at"),
        Index("idx_action_checkin_action_id", "action_id"),
    )


class ActionStreak(ContractAuditMixin, Base):
    """行动连续打卡统计（t_action_streak）。

    streak_status：ACTIVE / BROKEN / NEVER。
    """
    __tablename__ = "t_action_streak"

    id: Mapped[int] = mapped_column(BigIntPK, primary_key=True, autoincrement=True)
    user_id: Mapped[UUID] = mapped_column(GUID(), nullable=False)
    current_streak_days: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default=text("0")
    )
    longest_streak_days: Mapped[int] = mapped_column(
        Integer, nullable=False, default=0, server_default=text("0")
    )
    last_checkin_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    streak_status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="NEVER", server_default=text("'NEVER'")
    )
    # ACTIVE / BROKEN / NEVER

    __table_args__ = (
        UniqueConstraint("user_id", name="uk_action_streak_user_id"),
    )


class ActionWeight(ContractAuditMixin, Base):
    """行动类型权重配置（t_action_weight）。"""
    __tablename__ = "t_action_weight"

    id: Mapped[int] = mapped_column(BigIntPK, primary_key=True, autoincrement=True)
    action_type: Mapped[str] = mapped_column(String(20), nullable=False)
    weight: Mapped[int] = mapped_column(
        Integer, nullable=False, default=1, server_default=text("1")
    )
    weight_label: Mapped[str] = mapped_column(String(100), nullable=False)
    enabled: Mapped[bool] = mapped_column(
        Boolean, nullable=False, default=True, server_default=text("TRUE")
    )

    __table_args__ = (
        UniqueConstraint("action_type", name="uk_action_weight_action_type"),
    )
